File: Sample/Sample/PushSettingsViewController.m
/* Copyright 2018 Urban Airship and Contributors */

@import AirshipKit;
#import "PushSettingsViewController.h"

@interface PushSettingsViewController ()

@end

@implementation PushSettingsViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(refreshView) name:@"channelIDUpdated" object:nil];

    // Initialize switches
    self.locationEnabledSwitch.on = [UAirship location].locationUpdatesEnabled;
    self.analyticsSwitch.on = [UAirship analytics].enabled;

    // Add observer to didBecomeActive to update upon retrun from system settings screen
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didBecomeActive) name:@"UIApplicationDidBecomeActiveNotification" object:nil];

    self.locationEnabledLabel.text = NSLocalizedStringFromTable(@"UA_Location_Enabled", @"UAPushUI", @"Location Enabled label");
    self.locationEnabledSubtitleLabel.text = NSLocalizedStringFromTable(@"UA_Location_Enabled_Detail", @"UAPushUI", @"Enable GPS and WIFI Based Location detail label");
}

- (void)didBecomeActive {
    [self refreshView];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self refreshView];
}

- (IBAction)switchValueChanged:(id)sender {

    // Only allow disabling user notifications on iOS 10+
    if ([[NSProcessInfo processInfo] isOperatingSystemAtLeastVersion:(NSOperatingSystemVersion){10, 0, 0}]) {
        [UAirship push].userPushNotificationsEnabled = self.pushEnabledSwitch.on;
    } else if (self.pushEnabledSwitch.on) {
        [UAirship push].userPushNotificationsEnabled = YES;
    }

    [UAirship location].locationUpdatesEnabled = self.locationEnabledSwitch.on;

    [UAirship analytics].enabled = self.analyticsSwitch.on;
}

- (void)refreshView {
    self.pushEnabledSwitch.on = [UAirship push].userPushNotificationsEnabled;

    self.channelIDSubtitleLabel.text = [UAirship push].channelID;

    self.namedUserSubtitleLabel.text = [UAirship namedUser].identifier == nil ? NSLocalizedStringFromTable(@"None", @"UAPushUI", @"None") : [UAirship namedUser].identifier;

    if ([UAirship push].tags.count) {
        self.tagsSubtitleLabel.text = [[UAirship push].tags componentsJoinedByString:@", "];
    } else {
        self.tagsSubtitleLabel.text = NSLocalizedStringFromTable(@"None", @"UAPushUI", @"None");
    }
}

- (NSString *)pushTypeString {
    UAAuthorizedNotificationSettings authorizedSettings = [UAirship push].authorizedNotificationSettings;

    NSMutableArray *settingsArray = [NSMutableArray arrayWithCapacity:3];

    if (authorizedSettings & UAAuthorizedNotificationSettingsAlert) {
        [settingsArray addObject:NSLocalizedStringFromTable(@"UA_Notification_Type_Alerts", @"UAPushUI", @"Alerts")];
    }

    if (authorizedSettings & UAAuthorizedNotificationSettingsBadge) {
        [settingsArray addObject:NSLocalizedStringFromTable(@"UA_Notification_Type_Badges", @"UAPushUI", @"Badges")];
    }

    if (authorizedSettings & UAAuthorizedNotificationSettingsSound) {
        [settingsArray addObject:NSLocalizedStringFromTable(@"UA_Notification_Type_Sounds", @"UAPushUI", @"Sounds")];
    }

    if (![settingsArray count]) {
        return NSLocalizedStringFromTable(@"UA_Push_Settings_Link_Disabled_Title", @"UAPushUI", @"Pushes Currently Disabled");
    }

    return [settingsArray componentsJoinedByString:@", "];
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];

    if (indexPath.section == [tableView indexPathForCell:self.channelIDCell].section) {
        if (indexPath.row == [tableView indexPathForCell:self.channelIDCell].row) {
            if ([UAirship push].channelID) {
                [UIPasteboard generalPasteboard].string = self.channelIDSubtitleLabel.text;
                [self showCopyMessage];
            }
        }
    }
}

- (void)showCopyMessage {
    UIAlertController* alert = [UIAlertController alertControllerWithTitle:nil
                                                                   message:NSLocalizedStringFromTable(@"UA_Copied_To_Clipboard", @"UAPushUI", @"Copied to clipboard string")
                                                            preferredStyle:UIAlertControllerStyleAlert];

    UIAlertAction *okAction = [UIAlertAction actionWithTitle:NSLocalizedStringFromTable(@"UA_OK", @"UAPushUI", @"OK button string")
                                                       style:UIAlertActionStyleDefault
                                                     handler:^(UIAlertAction * _Nonnull action) {
        [self dismissViewControllerAnimated:YES completion:nil];
    }];

    [alert addAction:okAction];

    [self presentViewController:alert animated:YES completion:nil];
}

@end

File: SwiftSample/SwiftSample/PushSettingsViewController.swift
/* Copyright 2018 Urban Airship and Contributors */

import AirshipKit
import UIKit

class PushSettingsViewController: UITableViewController {

    @IBOutlet var pushEnabledCell: UITableViewCell!
    @IBOutlet var channelIDCell: UITableViewCell!
    @IBOutlet var namedUserCell: UITableViewCell!
    @IBOutlet var tagsCell: UITableViewCell!
    @IBOutlet var locationEnabledCell: UITableViewCell!

    @IBOutlet var pushEnabledSwitch: UISwitch!
    @IBOutlet var locationEnabledSwitch: UISwitch!
    @IBOutlet var analyticsSwitch: UISwitch!

    @IBOutlet var pushSettingsLabel: UILabel!
    @IBOutlet var pushSettingsSubtitleLabel: UILabel!
    @IBOutlet var locationEnabledLabel: UILabel!
    @IBOutlet var locationEnabledSubtitleLabel: UILabel!
    @IBOutlet var channelIDSubtitleLabel: UILabel!
    @IBOutlet var namedUserSubtitleLabel: UILabel!
    @IBOutlet var tagsSubtitleLabel: UILabel!

    @IBAction func switchValueChanged(_ sender: UISwitch) {

        // Only allow disabling user notifications on iOS 10+
        if (ProcessInfo().isOperatingSystemAtLeast(OperatingSystemVersion(majorVersion: 10, minorVersion: 0, patchVersion: 0))) {
            UAirship.push().userPushNotificationsEnabled = pushEnabledSwitch.isOn
        } else if (pushEnabledSwitch.isOn) {
            UAirship.push().userPushNotificationsEnabled = true
        }

        UAirship.location().isLocationUpdatesEnabled = locationEnabledSwitch.isOn
        UAirship.shared().analytics.isEnabled = analyticsSwitch.isOn
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(PushSettingsViewController.refreshView),
            name: NSNotification.Name("channelIDUpdated"),
            object: nil);

        // Initialize switches
        locationEnabledSwitch.isOn = UAirship.location().isLocationUpdatesEnabled
        analyticsSwitch.isOn = UAirship.shared().analytics.isEnabled

        // add observer to didBecomeActive to update upon retrun from system settings screen
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(PushSettingsViewController.didBecomeActive),
            name: NSNotification.Name.UIApplicationDidBecomeActive,
            object: nil)

        locationEnabledLabel.text = NSLocalizedString("UA_Location_Enabled", tableName: "UAPushUI", comment: "Location Enabled label")
        locationEnabledSubtitleLabel.text = NSLocalizedString("UA_Location_Enabled_Detail", tableName: "UAPushUI", comment: "Enable GPS and WIFI Based Location detail label")
    }

    // this is necessary to update the view when returning from the system settings screen
    @objc func didBecomeActive () {
        refreshView()
    }

    override func viewWillAppear(_ animated: Bool) {
        refreshView()
    }

    @objc func refreshView() {
        pushEnabledSwitch.isOn = UAirship.push().userPushNotificationsEnabled

        channelIDSubtitleLabel?.text = UAirship.push().channelID
        
        namedUserSubtitleLabel?.text = UAirship.namedUser().identifier == nil ? NSLocalizedString("None", tableName: "UAPushUI", comment: "None") : UAirship.namedUser().identifier

        if (UAirship.push().tags.count > 0) {
            self.tagsSubtitleLabel?.text = UAirship.push().tags.joined(separator: ", ")
        } else {
            self.tagsSubtitleLabel?.text = NSLocalizedString("None", tableName: "UAPushUI", comment: "None")
        }
    }

    func pushTypeString () -> String {

        let authorizedSettings = UAirship.push().authorizedNotificationSettings;

        var settingsArray: [String] = []

        if (authorizedSettings.contains(.alert)) {
            settingsArray.append(NSLocalizedString("UA_Notification_Type_Alerts", tableName: "UAPushUI", comment: "Alerts"))
        }

        if (authorizedSettings.contains(.badge)){
            settingsArray.append(NSLocalizedString("UA_Notification_Type_Badges", tableName: "UAPushUI", comment: "Badges"))
        }

        if (authorizedSettings.contains(.sound)) {
            settingsArray.append(NSLocalizedString("UA_Notification_Type_Sounds", tableName: "UAPushUI", comment: "Sounds"))
        }

        if (settingsArray.count == 0) {
            return NSLocalizedString("UA_Push_Settings_Link_Disabled_Title", tableName: "UAPushUI", comment: "Pushes Currently Disabled")
        }

        return settingsArray.joined(separator: ", ")
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)

        if ((indexPath as NSIndexPath).section == (tableView.indexPath(for: channelIDCell)! as NSIndexPath).section &&
            ((indexPath as NSIndexPath).row) == (tableView.indexPath(for: channelIDCell)! as NSIndexPath).row) {
                if ((UAirship.push().channelID) != nil) {
                    UIPasteboard.general.string = channelIDSubtitleLabel?.text
                    showCopyMessage()
                }
        }
    }
    
    func showCopyMessage () {
        let message = NSLocalizedString("UA_Copied_To_Clipboard", tableName: "UAPushUI", comment: "Copied to clipboard string")

        let alert = UIAlertController(title: nil, message: message, preferredStyle: UIAlertControllerStyle.alert)
        let buttonTitle = NSLocalizedString("UA_OK", tableName: "UAPushUI", comment: "OK button string")

        let okAction = UIAlertAction(title: buttonTitle, style: UIAlertActionStyle.default, handler: { (action: UIAlertAction) in
            self.dismiss(animated: true, completion: nil)
        })

        alert.addAction(okAction)

        self.present(alert, animated: true, completion: nil)
    }
}




